include/zephyr/bluetooth/gatt.h

```diff
@@ -1817,6 +1828,39 @@ int bt_gatt_unsubscribe(struct bt_conn *conn,
  */
 void bt_gatt_cancel(struct bt_conn *conn, void *params);
 
+/** @brief Encode GATT UUID of current services into a bt_data struct.
+ *
+ * The data will include both primary and secondary service UUIDs.
+ * Mandatory services such as the GATT and GAP services will not be included.
+ *
+ * @param[in] type
+ * Type of the UUIDs. Shall be one of following:
+ *  - BT_DATA_UUID16_SOME:  Encode 16-bit UUIDs until @p data_buf is full.
+ *  - BT_DATA_UUID16_ALL:   Encode all 16-bit UUIDs, and return error if they
+ *                          do not fit in @p data_buf.
+ *  - BT_DATA_UUID32_SOME:  Encode 32-bit UUIDs until @p data_buf is full.
+ *  - BT_DATA_UUID32_ALL:   Encode all 32-bit UUIDs, and return error if they
+ *                          do not fit in @p data_buf.
+ *  - BT_DATA_UUID128_SOME: Encode 128-bit UUIDs until @p data_buf is full.
+ *  - BT_DATA_UUID128_ALL:  Encode all 128-bit UUIDs, and return error if they
+ *                          do not fit in @p data_buf.
+ * @param[out] data           Pointer to the bt_data struct to store the
+ *                            results in.
+ * @param[out] data_buf       Pointer to the buffer where the encoded UUID data
+ *                            is stored.
+ * @param[in] data_buf_size   Size of the @p data_buf
+ *
+ * @retval 0 on success.
+ *
+ * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
+ *         In this case @p data_buf and @p data will be invalid.
+ *
+ * @retval -ENOMEM if a _ALL type is used and @p data_buf does not fit all the
+ *         data. In this case @p data_buf and @p data will be invalid.
+ */
+int bt_gatt_get_svc_uuid_data(uint8_t type, struct bt_data *data,
+			      uint8_t data_buf[], size_t data_buf_size);
```

Comment:
Couldn't this API be simplified to have just two input parameters, if you require the caller to initialize the `bt_data` struct up-front with the type and buf values? Then the only additional thing you need to provide is the maximum length (i.e. size of the buffer).

You also either need to fix the documentation or the API name to use the same verb. In the documentation you say "encode" but in the function name you say "get".

Reply:
Just to add to this, if you insist on keeping the `type` and `data_buf` separate and only assigned to `bt_data` inside the call, then I would at least order the parameters so that `bt_data` becomes the first parameter,  Since that's really the only "out" parameter (I'm not sure it's right to call `data_buf` "out", since I don't think the caller is supposed to directly touch its content after the call, rather the "end product" of the call is the `bt_data` struct.

Reply:
I agree with your comments, and we can require that the user initializes the `bt_data` first with the `type` and `data` pointer. 

I'll try to make this change in a new commit (so it's easier to revert), and then I can just squash it if that's what we want.

Reply:
@jhedberg Should I revert the commit to my original proposed API, or should we do something else? 

We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

Reply:
Hmm, that's a tricky one. The reason it's const is to allow assignments to it from runtime evaluated literal values (like most samples are now doing).

> We could also, instead of using `struct bt_data`, simply put all UUID data in a `uint8_t` buffer, and then let the let the user put that buffer into their `struct bt_data`.

Yeah, that might be a cleaner solution, although the encoding format really only has a single user (advertising data). That said, the format is not special - just an array of consecutive values.

Reply:
Btw, if we go with this "encode into uint8_t array" solution  we also need to let the caller know if everything fit or not, and then let the caller then decide the advertising data type value. Or to put it in other words, when we remove the bt_data type from the API, it should not use BT_DATA_UUID* values as either input or output, to make it generic in clean way.

Reply:
...it might be possible to convince me otherwise on that last point.. just thinking out loud. Let me know what you think

Reply:
I'm also just thinking aloud here:

```c
 * @retval Number of octets put in @p data_buf.
 *
 * @retval -EINVAL if any of the parameters are invalid (e.g. NULL or 0).
 *         In this case @p data_buf and @p data will be invalid.
 *
 * @retval -ENOMEM if not all service UUIDs could be placed in @p data_buf. 
 *         @p data_buf will contain as many UUIDs as could be placed. In this 
 *         case @p data_buf can still be used with the BT_DATA_UUID16_SOME, 
 *         BT_DATA_UUID32_SOME or BT_DATA_UUID128_SOME depending on the @p type.
 */
int bt_gatt_get_svc_uuid_data(uint8_t type, uint8_t data_buf, size_t data_buf_size);
```

and where `type` would be one of BT_UUID_TYPE_16, BT_UUID_TYPE_32, BT_UUID_TYPE_128.

i.e. the function will pack as many service UUIDs in the `data_buf` as possible, and return the number of octets put in it. 

The returned buffer and size should be enough to then encode the `struct bt_data`. 

But like others, I do have doubt about the usefulness of this API, as A) it's pretty trivial to encode the service UUIDs yourself today, and it's quite difficult to provide a good API for this.

Reply:
@jhedberg @MariuszSkamra What do you think? Is it worth pursuing this API, or just leave it?

Reply:
I'm ok leaving it. We can always revisit it later in the future if necessary (e.g. new use cases or users come along)

Reply:
@jhedberg Should we leave it open, move it to draft or close it then? 

If we are not planning on merging it anytime soon, I'm in favor of just closing it.

Reply:
Closing is probably the best option, but maybe keep the branch in your fork so that this can be revisited later.